src/ft60x.rs

```diff
@@ -97,37 +105,131 @@ impl FT60x {
+                .chunks_mut(blocksize)
+                {
+                    // The FT60x doesn't seem to like too many outstanding requests
+                    if outstanding > 500 {
+                        let mut to_ship = None;
+                        for (i, async_group) in async_group_buffer
+                            .iter_mut()
+                            .map(|v| &mut v.1)
+                            .chain(once(&mut current_async_group))
+                            .enumerate()
+                        {
+                            match async_group.wait_any() {
+                                Ok(mut transfer) => {
+                                    ensure!(
+                                        transfer.buffer().len() == transfer.actual().len(),
+                                        "FT60x did not return enough data. requested {} got {}",
+                                        transfer.buffer().len(),
+                                        transfer.actual().len()
+                                    );
+                                    outstanding -= 1;
+                                    break;
+                                }
+                                Err(rusb::Error::NotFound) => {
+                                    assert_eq!(to_ship, None);
+                                    assert_eq!(i, 0);
+                                    to_ship = Some(i);
+                                }
+                                Err(_) => panic!(),
+                            }
+                        }
+
+                        if let Some(i) = to_ship {
+                            if i < async_group_buffer.len() {
+                                full_buffer_tx.send(async_group_buffer.remove(i).0).unwrap();
+                            }
+                        }
+                    }
+
+                    current_async_group.submit(Transfer::bulk(
+                        &self.device,
+                        0x82,
+                        chunk,
+                        Duration::new(1, 0),
+                    ));
+                    outstanding += 1;
+                }
+                async_group_buffer.push((current_buffer, current_async_group));
+            }
+
+            Ok(())
+        };
+
+        thread::Builder::new()
+            .name("ft60x-rx".to_string())
+            .spawn(move || {
+                thread_fn().unwrap();
+            });
+
+        (empty_buffer_tx, full_buffer_rx)
+    }
+
+    /// it is recommended to request multiples of 32Kb
+    pub fn data_stream_ringbuf(mut self, bufsize: usize) -> Result<RingBufConsumer<Vec<u8>>> {
         let (mut producer, consumer) =
             RingBuf::<Vec<u8>>::create_channel_with_default_value(4, vec![0u8; bufsize]);
 
         std::thread::spawn(move || {
             while producer
-                .with_next_buffer(|buf| self.read_exact(buf))
+                .with_next_buffer(|buf| self.read_exact(buf).unwrap())
```

Comment:
would be nice to forward errors from within ` with_next_buffer` to the outside world, but maybe that can wait for a future PR :)